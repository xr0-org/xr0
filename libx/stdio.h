#ifndef STDIO_H
#define STDIO_H

#include <stddef.h>

typedef int FILE;

axiom int
putc(int c);

axiom int
putchar(int c);

axiom int
fputs(char *s, FILE *stream);

axiom int
puts(char *s);

axiom FILE *
fopen(char *pathname, char *mode);

axiom int
fseek(FILE *stream, int offset, int whence); /* TODO: convert offset to long */

#define SEEK_END 0
#define SEEK_SET 0

axiom int
ftell(FILE *stream);

axiom size_t
fread(void *ptr, size_t size, size_t nmemb, FILE *stream);

axiom int
fclose(FILE *stream);

/* XXX: provisionally rigid for preconditions example */
axiom char *
sprintf(char *buf, char *format, int *arg) ~ [
	*buf = $;
];

/* XXX: provisionally rigid for preconditions example */
axiom int
scanf(char *format, int *arg) ~ [
	/* since there's a sideeffect here for arg, we need to characterise it */
	*arg = $;
];

#endif
