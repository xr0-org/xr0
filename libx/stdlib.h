#ifndef STDLIB_H
#define STDLIB_H

axiom void *
malloc(int size) [ .alloc result; ];

axiom void
free(void *ptr) [ .dealloc ptr; ];

#endif
