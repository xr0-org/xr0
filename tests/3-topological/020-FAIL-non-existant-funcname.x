#include <stdlib.h>
#include <stdio.h>

void *
main2()
{
	return NULL;
}
