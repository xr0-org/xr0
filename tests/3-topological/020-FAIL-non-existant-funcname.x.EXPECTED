function `main' is not declared
